features/26-mfa/api.md

```diff
@@ -0,0 +1,436 @@
+  #
+  # 'required' means the user must login with MFA.
+  # APIs like change password will return MFA required error.
+  # This is like how banking services offer MFA.
+  #
+  # Default is 'optional'
+  enforcement: 'optional'
+  bearer_token:
+    # How many days the bearer token is valid. Default is 30.
+    expire_in_days: 30
+  recovery_code:
+    # How many recovery code should be generated. Default is 16.
+    count: 16
+```
+
+# SDK API
+
+The following functions are all in the namespace `skygear.auth.mfa`.
+
+```typescript
+// Register TOTP
+
+interface CreateNewTOTPResult {
+  authenticatorID: string;
+  authenticatorType: "totp";
+  secret: string;
+}
+
+interface ActivateTOTPResult {
+  recoveryCodes?: string[];
+}
+
+async function createNewTOTP(displayName?: string): Promise<CreateNewTOTPResult>;
+async function activateTOTP(authenticatorID: string, otp: string): Promise<ActivateTOTPResult>;
+
+// Register OOB
+
+interface CreateNewOOBOptions {
+  channel?: "sms";
+  phone: string;
+}
+
+interface CreateNewOOBResult {
+  authenticatorID: string;
+  authenticatorType: "oob";
+  channel: "sms";
+  phone: string;
+}
+
+interface ActivateOOBResult {
+  recoveryCodes?: string[];
+}
+
+async function createNewOOB(options: CreateNewOOBOptions): Promise<CreateNewOOBResult>;
+async function triggerOOB(authenticatorID?: string): Promise<void>;
+async function activateOOB(authenticatorID: string, code: string): Promise<ActivateOOBResult>;
+
+// Error inspection
+
+function isMFARequiredError(err: unknown): boolean;
+
+// Authenticate
+
+interface AuthenticateExtraOptions {
+  requestBearerToken?: boolean;
+}
+
+interface AuthenticateResult {
+  bearerToken?: string;
+}
+
+interface TOTPAuthenticateOptions {
+  authenticatorID?: string;
+  otp: string;
+}
+
+async function authenticateWithTOTP(options: TOTPAuthenticateOptions, extraOptions?: AuthenticateExtraOptions): Promise<AuthenticateResult>;
+
+interface OOBAuthenticateOptions {
+  authenticatorID?: string;
+  code: string;
+}
+
+async function authenticateWithOOB(options: OOBAuthenticateOptions, extraOptions?: AuthenticateExtraOptions): Promise<AuthenticateResult>;
+
+async function authenticateWithRecoveryCode(code: string, extraOptions?: AuthenticateExtraOptions): Promise<AuthenticateResult>;
+
+async function authenticateWithBearerToken(token: string): Promise<void>;
+
+// Authenticator management
+
+type Authenticator = TOTPAuthenticator | OOBAuthenticator | RecoveryCodeAuthenticator | BearerTokenAuthenticator;
+
+interface TOTPAuthenticator {
+  id: string;
+  type: "totp";
+  activatedAt: Date;
+  displayName: string;
+}
+
+interface OOBAuthenticator {
+  id: string;
+  type: "oob";
+  activatedAt: Date;
+  channel: "sms";
+  phone: string;
+}
+
+interface RecoveryCodeAuthenticator {
+  id: string;
+  type: "recovery_code";
+  createdAt: Date;
+}
+
+interface BearerTokenAuthenticator {
+  id: string;
+  type: "bearer_token";
```

Comment:
I'm not sure if bearer token should be a standalone authenticator: if I login with a TOTP authenticator and get a token, the token should be invalidated if the TOTP authenticator is deleted.

Reply:
Good point! Maybe for bearer token authenticator it must have a reference to an existing authenticator (only "totp", "oob" and "recovery_code", does not make sense to depend on another bearer token authenticator). When the parent authenticator is deleted, the dependent bearer token authenticator(s) must also be deleted.

Reply:
I think it does not make sense to depends on a recovery code authenticator too?